src/gensxg.c

```diff
@@ -1,128 +1,217 @@
+  const char* content_type;
+  sxg_header_t header;
+  uint64_t mi_record_size;
   const char* url;
   const char* cert_url;
   const char* validity_url;
   const char* certificate;
   const char* public_key;
   const char* private_key;
-  const char* content;
-  const char* content_type;
-  uint64_t date;
-  uint64_t expire;
-  uint64_t mi_record_size;
+  time_t date;
+  time_t expire;
 } Options;
 
 static EVP_PKEY* load_private_key(const char* filepath) {
   FILE* const keyfile = fopen(filepath, "r");
   if (keyfile == NULL) {
-    return NULL;
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
   }
   EVP_PKEY* private_key =
       PEM_read_PrivateKey(keyfile, NULL, NULL, NULL);
+  if (private_key == NULL) {
+    fprintf(stderr, "Failed to read private key: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
   fclose(keyfile);
   return private_key;
 }
 
 static X509* load_x509_cert(const char* filepath) {
   FILE* certfile = fopen(filepath, "r");
   if (certfile == NULL) {
-    return NULL;
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
   }
   X509* cert = PEM_read_X509(certfile, NULL, NULL, NULL);
+  if (cert == NULL) {
+    fprintf(stderr, "Failed to read cetificate: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
   fclose(certfile);
   return cert;
 }
 
 static EVP_PKEY* load_ed25519_pubkey(const char* filepath) {
   FILE* keyfile = fopen(filepath, "r");
-  if (!keyfile) {
-    return NULL;
+  if (keyfile == NULL) {
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
   }
   EVP_PKEY* public_key = PEM_read_PUBKEY(keyfile, NULL, NULL, NULL);
+  if (public_key == NULL) {
+    fprintf(stderr, "Failed to read public key: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
   fclose(keyfile);
   return public_key;
 }
 
-Options parse_options(int argc, char* const argv[]) {
+static time_t parse_time(const char datestring[]) {
+  struct tm date;
+  char* finished = strptime(datestring, "%Y-%m-%dT%H:%M:%SZ", &date);
+  if (finished == NULL) {
+    fprintf(stderr, "Failed to parse date: %s\n", datestring);
+  } else if(*finished != '\0') {
+    fprintf(stderr, "Failed to parse at: %s\n", finished);
+    exit(EXIT_FAILURE);
+  }
+  return mktime(&date);
+}
+
+Options init_default_options() {
   Options result;
   memset(&result, 0, sizeof(result));
-  // Set default values.
   result.help = false;
+  result.integrity_mode = false;
+  result.content = "index.html";
   result.content_type = "text/html";
+  result.header = sxg_empty_header();
   time_t now = time(NULL);
-  result.date = (uint64_t)now;
-  result.expire = result.date + 60 * 60 * 24 * 7;  // 7 days.
+  result.date = now;
   result.mi_record_size = 4096;
+  return result;
+}
 
+static void header_append_string(const char* data, sxg_header_t* header) {
+  size_t colon_pos = 0;
+  for (size_t i = 0; data[i] != '\0'; ++i) {
+    if (colon_pos == 0 && data[i] == ':') {
```

Comment:
Personally I prefer `if (data[i] == ':') { colon_pos = i; break; }`, which is faster and simpler.

Reply:
Done.